### PR TITLE
:sparkles: update hooks create & delete path

### PR DESCRIPTION
**Pull request by flo-sch**
- read core.hooksPath from git config (if set)
- default to "process.cwd()/.git/hooks"
- adjust unit tests

## Description

Change how the path for git hooks is retrieved:
It now invokes `git config --get core.hooksPath` which provides a *relative* path (default to `.git/hooks` if unset), then create/delete the hooks at that path.

NOTE: @carloscuesta, the previous `git rev-parse` command returned an absolute folder.
Not sure if it is worth defaulting to that command, in case `core.hooksPath` is not set? (to be fully 100% backward-compatible?)

(I do not know if people would try to set-up gitmoji outside of the current working directory, feels edge-case to me 🤷‍♂️)

Issue: #577

## Tests

- [x] All tests passed.

### CODE REVIEW

File: src/utils/isHookCreated.js [hunk 1]
```diff
@@ -1,13 +1,14 @@
...
 import fs from 'fs'
+import path from 'path'
 
 import HOOK from '../commands/hook/hook'
 
 const isHookCreated = async () => {
   try {
-    const { stdout } = await execa('git', ['rev-parse', '--absolute-git-dir'])
-    const hookFile = stdout + HOOK.PATH
+    const { stdout } = await execa('git', ['config', '--get', 'core.hooksPath'])
```
Comment (carloscuesta) [inline src/utils/isHookCreated.js:10]: Shouldn't we check here for the default hooks path if no core.hooksPath is provided?

**Review by carloscuesta — COMMENTED** [review]
Thanks for opening up a PR ❤️ 

Great work! A couple of details:

File: src/commands/hook/create/index.js [hunk 2]
```diff
@@ -2,17 +2,30 @@
...
+ * If "core.hooksPath" is not set in git config, use default ".git/hooks"
+ * @see https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks
+ */
 const createHook = async () => {
   const spinner = ora('Creating the gitmoji commit hook').start()
 
   try {
-    const { stdout } = await execa('git', ['rev-parse', '--absolute-git-dir'])
+    const { stdout } = await execa('git', ['config', '--get', 'core.hooksPath'])
+    const relativeHooksPath = stdout || '.git/hooks'
```
[the remark this answers is not in the document]
Comment (flo-sch) [reply src/commands/hook/create/index.js:20]: That would potentially get me:

1. `/Users/foo/projects/bar/.git`
2. `.git/hooks`

Should I then merge those by replacing the common `.git` do you think?

[the remark this answers is not in the document]
Comment (flo-sch) [reply src/commands/hook/create/index.js:20]: Nevermind, according to the doc, the `core.hooksPath` in config could be either absolute or relative actually...
I will come with a new version hopefully covering all cases!

**Comment by flo-sch** [comment]
Alright, I forced pushed after signing my commits, I use 2 emails for my Git accounts and I often forget to switch to my personal one when contributing to OSS 😅 

Anyway, I have created a new helper to retrieve the hooks path following different scenarios:

1. If the hook path in the config is absolute, put the hook in there (apparently possible according to git documentation)
ex: `/etc/git/hooks`
2. If it is relative, get the absolute path of the git directory, use `dirname()` to remove the `.git` part, then apply the hooks path (since according to my understanding, it is relative to the root folder)
ex: `/Users/foo/projects/bar/.git/hooks` or `/Users/carloscuesta/GitHub/gitmoji-cli/.husky`

File: src/commands/hook/remove/index.js [hunk 3]
```diff
@@ -1,23 +1,20 @@
...
-    const { stdout } = await execa('git', ['rev-parse', '--absolute-git-dir'])
+    const hookFile = await getAbsoluteHooksPath(HOOK.FILENAME)
 
-    fs.unlink(stdout + HOOK.PATH, (error) => {
-      if (error)
-        return spinner.fail('Error: Gitmoji commit hook is not created')
-      spinner.succeed('Gitmoji commit hook removed successfully')
-    })
+    fs.unlinkSync(hookFile)
+    spinner.succeed('Gitmoji commit hook removed successfully')
```
Comment (carloscuesta) [inline src/commands/hook/remove/index.js:15]: I'm curious, what did we changed this? 🤔

File: src/utils/getAbsoluteHooksPath.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+      // Use dirname to remove the trailing "/.git" from the path
+      path.dirname(absoluteGitDirCommand.stdout),
+      hooksPath
+    )
+  }
+
+  return path.resolve(absoluteHooksPath, hookName)
+}
+
+export default getAbsoluteHooksPath
```
Comment (carloscuesta) [inline src/utils/getAbsoluteHooksPath.js:40]: I don't understand why we need to check if it's absolute or not, can we make something like this?

```suggestion
// @flow
import execa from 'execa'
import path from 'path'

const getGitHooksPath = async (hookName: string): string => {
  const gitCoreHooksPath = await execa('git', [
    'config',
    '--get',
    'core.hooksPath'
  ])

  const defaultHooksPath = await execa('git', [
    'rev-parse',
    '--absolute-git-dir'
  ]) + '/hooks'
  
  const hooksPath = gitCoreHooksPath.stdout || defaultHooksPath.stdout

  return path.resolve(hooksPath, hookName)
}

export default getGitHooksPath
```

File: test/commands/hook.spec.js [hunk 5]
```diff
@@ -15,46 +16,62 @@ describe('hook command', () => {
   })
 
   describe('create hook', () => {
-    beforeAll(() => {
-      execa.mockReturnValue({ stdout: stubs.gitAbsoluteDir })
-      hook.create()
+    beforeEach(() => {
+      execa.mockReturnValueOnce({ stdout: stubs.relativeCoreHooksPath })
+      execa.mockReturnValueOnce({ stdout: stubs.gitAbsoluteDir })
```
Comment (carloscuesta) [inline test/commands/hook.spec.js:21]: I think that here we're testing the internal implementation of `getAbsoluteHooksPath`, in this spec I would suggest to mock this module and test the implementation on the `getAbsoluteHooksPath` spec

File: test/setupTests.js [hunk 6]
```diff
@@ -5,7 +5,9 @@ jest.mock('path-exists')
 jest.mock('fs')
 jest.mock('ora', () =>
   jest.fn().mockReturnValue({
-    start: jest.fn(),
+    start: jest.fn(function () {
+      return this
+    }),
```
Comment (carloscuesta) [inline test/setupTests.js:10]: 🤔

File: test/utils/getAbsoluteHooksPath.spec.js [hunk 7]
```diff
@@ -0,0 +1,89 @@
+import execa from 'execa'
+// import fs from 'fs'
+import path from 'path'
+
+import getAbsoluteHooksPath from '../../src/utils/getAbsoluteHooksPath'
+import * as stubs from './stubs'
+
+describe('getAbsoluteHooksPath', () => {
```
Comment (carloscuesta) [inline test/utils/getAbsoluteHooksPath.spec.js:8]: Maybe we can simplify this spec if we refactor the utility!

File: test/utils/isHookCreated.spec.js [hunk 8]
```diff
@@ -1,13 +1,15 @@
...
 import isHookCreated from '../../src/utils/isHookCreated'
 import HOOK from '../../src/commands/hook/hook'
 import * as stubs from './stubs'
 
 describe('isHookCreated', () => {
-  beforeAll(() => {
-    execa.mockReturnValue({ stdout: stubs.gitAbsoluteDir })
+  beforeEach(() => {
+    execa.mockReturnValueOnce({ stdout: stubs.relativeCoreHooksPath })
+    execa.mockReturnValueOnce({ stdout: stubs.gitAbsoluteDir })
```
Comment (carloscuesta) [inline test/utils/isHookCreated.spec.js:12]: The same as before, we're testing the internal implementation of the util, let's mock the util in this spec!

**Review by carloscuesta — COMMENTED** [review]
Thanks!

File: src/utils/getAbsoluteHooksPath.js (hunk 4 above)
Comment (flo-sch) [reply src/utils/getAbsoluteHooksPath.js:40]: Somehow from your previous review, I thought you wanted to always ensure working with an absolute path, but I guess using just a relative path as you illustrate it works too :)

We would get the following results, based on the value of `core.hooksPath`

- `/etc/git/hooks` --> `/etc/git/hooks/prepare-commit-msg`
- `undefined` --> `/path/to/project/.git/hooks/prepare-commit-msg`
- `.git/hooks` --> `.git/hooks/prepare-commit-msg`
- `.husky` --> `.husky/prepare-commit-msg`

When my proposal would instead always return an absolute path:

- `/etc/git/hooks` --> `/etc/git/hooks/prepare-commit-msg`
- `undefined` --> `/path/to/project/.git/hooks/prepare-commit-msg`
- `.git/hooks` --> `/path/to/project/.git/hooks/prepare-commit-msg`
- `.husky` --> `/path/to/project/.husky/prepare-commit-msg`

File: src/commands/hook/remove/index.js (hunk 3 above)
Comment (flo-sch) [reply src/commands/hook/remove/index.js:15]: There were some `UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'fail' of undefined` in the tests, I felt easier to handle that with the `*Sync` API inside an async function

File: test/setupTests.js (hunk 6 above)
Comment (flo-sch) [reply test/setupTests.js:10]: Fix for: `UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'fail' of undefined`

The way spinners are used inside the hooks was by chaining the constructor and start method.
Ora's [start()](https://github.com/sindresorhus/ora#starttext) method returns its instance, allowing to then invoke other methods such as `succeed()` or `fail()`.
But if the mocked implementation of `start()` does not return its instance, then`spinner.fail()` or `spinner.succeed()` fails with the error message above

File: test/utils/getAbsoluteHooksPath.spec.js (hunk 7 above)
Comment (flo-sch) [reply test/utils/getAbsoluteHooksPath.spec.js:6]: Simpler version of the helper available, I kept the 3 same scenarios though, just simplified a bit the matchers.
Not sure if you want me to rewrite some?

**Review by carloscuesta — APPROVED** [review]
Awesome thanks for the work! 🚀
